### DISCUSSION TITLE
Can't use @unique key in `where` on connectOrCreate

### DISCUSSION THREAD

**Discussion by himynameisdave** [prisma/prisma#4382] [Q&A] [answered]
Hey there, I'm just getting up and running with Prisma and am trying to build a mock music review blog. So far things are mostly going great (the tooling is nice) but I am having trouble with `connectOrCreate`.

Specifically, I am trying to create a new entity, which may or may not exist in the db already. At the time I run `connectOrCreate` I do not know the primary key (`id`) for that entity, but I do have other unique fields that I would like to use in the `where` clause. Not sure why this isn't working:

`schema.prisma`

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Artist {
  id          String    @default(uuid()) @id
  name        String    @unique
  albums      Album[]
  createdAt   DateTime  @default(now())
  updatedAt   DateTime  @updatedAt
}

model Album {
  id          String    @default(uuid()) @id
  title       String
  albumArtURL String
  year        Int
  artist      Artist
  genres      Genre[]   @relation(references: [id])
  score       Float
  createdAt   DateTime  @default(now())
  updatedAt   DateTime  @updatedAt
}

model Genre {
  id          String    @default(uuid()) @id
  name        String    @unique
  albums      Album[]   @relation(references: [id])
}

model Review {
  id               String      @default(uuid()) @id
  url              String      @unique
  date             DateTime
  artist           Artist
  albums           Album[]
  text             ReviewText  @relation(fields: [textId], references: [id])
  textId           String
  author           Author
  createdAt        DateTime    @default(now())
  updatedAt        DateTime    @updatedAt
}

model ReviewText {
  id               String    @default(uuid()) @id
  abstract         String
  text             String
}

model Author {
  id               String    @default(uuid()) @id
  name             String    @unique
  title            String
  url              String?
  createdAt        DateTime  @default(now())
  updatedAt        DateTime  @updatedAt
}
```

`app.ts`

```typescript
const newReview = await prisma.review.create({
    data: {
      url: 'https://music.blog/reviews/good-album',
      date: new Date('November 24 2020'),
      text: {
        create: {
          abstract: 'This is the subheading.',
          text: 'This was a good album...',
        },
      },
      artist: {
        connectOrCreate: {
          where: {
            // ⚠️ This doesn't work, only lets me `where` on the `id` field.
            name: 'Some Artist Name',
          },
          create: {
            name: 'Some Artist Name',
          },
        },
      },
    },
  });
```

I thought that by assigning `@unique` attribute to the `name` field in `Artist` (as well as `Author`), I would be able to `where` off of them, but maybe I'm doing something wrong here?

Cheers!

---

MacOS Version: `10.14.6`
`@prisma/cli`/`@prisma/client` Version: `2.12.1`
Node Version: `v14.15.0`
Postgres Version: `13`

**Comment by ryands17** [comment, marked as answer]
@himynameisdave 👋 

Could you run `prisma generate` and try again? I have tried the above query and it works fine.
![image](https://user-images.githubusercontent.com/19697099/100435463-fad5be80-30c3-11eb-88b5-5d4b53562344.png)

`connectOrCreate` will always accept unique fields.

**Reply by himynameisdave** [reply]
Yep this worked like a charm, thanks for your quick reply. I didn't realize that I needed to run `generate` after a new migration, but it totally makes sense why those are separate.

Prisma is amazing! 😍 🎉

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
